geo/src/algorithm/line_intersection.rs

```diff
@@ -0,0 +1,666 @@
+    }
+    if p1q1p2 && q1p1q2 {
+        if q.start == p.start && !p1q2p2 && !q1p2q2 {
+            return Some(LineIntersection::SinglePoint {
+                intersection: q.start,
+                is_proper: false,
+            });
+        } else {
+            return Some(LineIntersection::Collinear {
+                intersection: Line::new(q.start, p.start),
+            });
+        }
+    }
+    if p1q1p2 && q1p2q2 {
+        if q.start == p.end && !p1q2p2 && !q1p1q2 {
+            return Some(LineIntersection::SinglePoint {
+                intersection: q.start,
+                is_proper: false,
+            });
+        } else {
+            return Some(LineIntersection::Collinear {
+                intersection: Line::new(q.start, p.end),
+            });
+        }
+    }
+    if p1q2p2 && q1p1q2 {
+        if q.end == p.start && !p1q1p2 && !q1p2q2 {
+            return Some(LineIntersection::SinglePoint {
+                intersection: q.end,
+                is_proper: false,
+            });
+        } else {
+            return Some(LineIntersection::Collinear {
+                intersection: Line::new(q.end, p.start),
+            });
+        }
+    }
+    if p1q2p2 && q1p2q2 {
+        if q.end == p.end && !p1q1p2 && !q1p1q2 {
+            return Some(LineIntersection::SinglePoint {
+                intersection: q.end,
+                is_proper: false,
+            });
+        } else {
+            return Some(LineIntersection::Collinear {
+                intersection: Line::new(q.end, p.end),
+            });
+        }
+    }
+    None
+}
+
+/// Finds the endpoint of the segments P and Q which is closest to the other segment.  This is
+/// a reasonable surrogate for the true intersection points in ill-conditioned cases (e.g.
+/// where two segments are nearly coincident, or where the endpoint of one segment lies almost
+/// on the other segment).
+///
+/// This replaces the older CentralEndpoint heuristic, which chose the wrong endpoint in some
+/// cases where the segments had very distinct slopes and one endpoint lay almost on the other
+/// segment.
+///
+/// `returns` the nearest endpoint to the other segment
+fn nearest_endpoint<F: GeoFloat>(p: Line<F>, q: Line<F>) -> Coordinate<F> {
+    use geo_types::private_utils::point_line_euclidean_distance;
+
+    let mut nearest_pt = p.start;
+    let mut min_dist = point_line_euclidean_distance(p.start, q);
+
+    let dist = point_line_euclidean_distance(p.end, q);
+    if dist < min_dist {
+        min_dist = dist;
+        nearest_pt = p.end;
+    }
+    let dist = point_line_euclidean_distance(q.start, p);
+    if dist < min_dist {
+        min_dist = dist;
+        nearest_pt = q.start;
+    }
+    let dist = point_line_euclidean_distance(q.end, p);
+    if dist < min_dist {
+        nearest_pt = q.end;
+    }
+    nearest_pt
+}
+
+fn raw_line_intersection<F: GeoFloat>(p: Line<F>, q: Line<F>) -> Option<Coordinate<F>> {
+    let min_x0 = if p.start.x < p.end.x {
```

Comment:
Is this the same as:

```rust
    let min_x0 = p.start.x.min(p.end.x);
```

https://docs.rs/num/0.4.0/num/trait.Float.html#tymethod.min

Reply:
Indeed it is! I've updated to use that, and it's much more concise.